plugins/builds/index.js

```diff
@@ -293,83 +391,329 @@ exports.register = (server, options, next) => {
+                    const buildConfig = {
+                        jobFactory,
+                        buildFactory,
+                        eventFactory,
+                        pipelineId,
+                        jobName: nextJobName,
+                        username,
+                        scmContext,
+                        build, // this is the parentBuild for the next build
+                        baseBranch: event.baseBranch || null
+                    };
+
+                    // Just start the build if falls in to these 2 scenarios
+                    // 1. No join
+                    // 2. ([~D,B,C]->A) currentJob=D, nextJob=A, joinList(A)=[B,C]
+                    //    joinList doesn't include C, so start A
+                    if (joinList.length === 0 || !joinListNames.includes(currentJobName)) {
+                        return createBuild(buildConfig);
+                    }
+
+                    return Promise.resolve().then(() => {
+                        if (!event.parentEventId) {
+                            return event.getBuilds();
+                        }
+
+                        // If parent event id, merge parent build status data and
+                        // rerun all builds in the path of the startFrom
+                        return eventFactory.get({ id: event.parentEventId })
+                            .then(parentEvent => parentEvent.getBuilds()
+                                .then(parentBuilds => removeDownstreamBuilds({
+                                    builds: parentBuilds,
+                                    startFrom: event.startFrom,
+                                    parentEvent
+                                }))
+                            )
+                            .then(upstreamBuilds => event.getBuilds()
+                                .then(builds => builds.concat(upstreamBuilds)));
+                    }).then(finishedBuilds => handleNextBuild({
+                        buildConfig,
+                        joinList,
+                        finishedBuilds,
+                        jobId: workflowGraph.nodes
+                            .find(node => node.name === trimJobName(nextJobName)).id
+                    }));
+                }));
+            });
+        }
+
+        // New implementation that allows external join
+        const pipelineFactory = server.root.app.pipelineFactory;
+        const event = await eventFactory.get({ id: build.eventId });
+        const workflowGraph = event.workflowGraph;
+        const nextJobs = workflowParser.getNextJobs(workflowGraph,
+            { trigger: currentJobName, chainPR: pipeline.chainPR });
+        // Create a join object like: {A:[B,C], D:[B,F]} where [B,C] join on A, [B,F] join on D, etc.
+        // This can include external jobs
+        const joinObj = nextJobs.reduce((obj, jobName) => {
+            obj[jobName] = workflowParser.getSrcForJoin(workflowGraph, { jobName });
```

Comment:
double check here for external build
`[~D, B,C] --> sd@123:A`
Then joinList of `sd@123:A` is `[sd@111:B, sd@111:C]`